### PR TITLE
Avax synpase stable

### PR DESCRIPTION
**Pull request by joshualyguessennd**
- Avax synapse StablePool

### CODE REVIEW

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+
+        if (data[0] > 0 || data[1] > 0 || data[2] > 0 || data[3] > 0) {
+            swapPool.addLiquidity(data, 0, now);
+        }
+        // deposit lpToken 
+        yvToken.deposit();
+        lastInvest = block.timestamp;
+    }
+
+    function prepareMigration(address _strategy) internal override {
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol:450]: Why did we remove ETH support?

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+        uint256 maxSlip = expectedOut.mul(DENOMINATOR.sub(slippageProtectionIn)).div(DENOMINATOR);
+
+        // invest all the tokens we have
+        uint256[] memory data = new uint256[](4);
+        data[0] = nusd.balanceOf(address(this));
+        data[1] = dai.balanceOf(address(this));
+        data[2] = usdc.balanceOf(address(this));
+        data[3] = usdt.balanceOf(address(this));
+
+        if (data[0] > 0 || data[1] > 0 || data[2] > 0 || data[3] > 0) {
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: This strategy can't assume what token will be deposited or in which pool we're gonna use it. It should be as generic as possible. Please refer to this [version](https://github.com/Tesseract-Finance/tesseract-strategies/blob/9357074271a7f37dfd20e2578d1d74224dcc7c95/synnapse/contracts/Strategy.sol#L385)

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+            uint256 _profit,
+            uint256 _loss,
+            uint256 _debtPayment
+        )
+    {
+        _debtPayment = _debtOutstanding;
+
+        uint256 debt = vault.strategies(address(this)).totalDebt;
+        uint256 currentValue = estimatedTotalAssets();
+        uint256 wantBalance = IERC20(targetToken).balanceOf(address(this));
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: We can't use target token for this, `want` should be used and should never be able to be changed

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+            //needed because of revert on priceperfullshare if 0
+            return 0;
+        }
+
+        uint256 pricePerShare = yvToken.pricePerShare();
+        //curve tokens are 1e18 decimals
+        return balance.mul(pricePerShare).div(1e18);
+    }
+
+    function nativeToWant(uint256 _amtInWei) public view override virtual returns (uint256) {}
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: This should be implemented

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+        }
+        
+        Strategy(newStrategy).initialize(_vault, _strategist, _rewards, _keeper,_yvToken, _swapPool, _poolSize, _maxSingleInvest, _minTimePerInvest, _slippageProtectionIn, _strategyName);
+
+        emit Cloned(newStrategy);
+    }
+
+
+    // These functions are useful for setting parameters of the strategy that may need to be adjusted.
+    // Set optimal token to sell harvested funds for depositing to Curve.
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: Remove unused code

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+            (, uint256 withdrawalLoss) = withdrawSome(toFree);
+            // when we withdraw we can lose money in the withdrawl
+            if (withdrawalLoss < _profit) {
+                _profit = _profit.sub(withdrawalLoss);
+            } else {
+                _loss = _loss.add(withdrawalLoss.sub(_profit));
+                _profit = 0;
+            }
+
+            wantBalance = IERC20(targetToken).balanceOf(address(this));
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: Same scenario as above

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+        address _swapPool,
+        address _yvToken,
+        string memory _strategyName
+    ) internal {
+        require(_poolSize > 1 && _poolSize < 5, "incorrect pool size");
+        require(want_decimals == 0, "Already Initialized");
+        
+        swapPool = ISwap(_swapPool);
+
+        if (swapPool.getToken(0) == want) {
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: Missing support for the native token of the chain (ETH for example)

**Review by pajicf — CHANGES_REQUESTED** [review]
Seems like a mix of the original strategy and some LP strategy dynamics which are simply not needed here.
The code on this [branch](https://github.com/Tesseract-Finance/tesseract-strategies/tree/polygon-synnapse/synnapse) should be mostly complete and not need too many changes. If I'm not mistaken here, i think it's gonna be easiest to reuse it and just adapt the tests

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+    uint8 private want_decimals;
+
+    uint8 public curveId;
+    address public targetToken;
+    bool public withdrawProtection;
+
+    bool internal forceHarvestTriggerOnce; // only set this to true externally when we want to trigger our keepers to harvest for us
+    uint256 public minHarvestCredit; // if we hit this amount of credit, harvest the strategy
+
+    IERC20 internal constant emissionToken = IERC20(0xCA87BF3ec55372D9540437d7a86a7750B42C02f4);
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: I think this is a confusing name, sound more like the token we'll receive as a reward. Maybe we should rename it to "lpToken" or something like that?

We also can't hard code it since this strategy should be able to be reused on other pools

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+
+    function nativeToWant(uint256 _amount) public view override virtual returns (uint256) {
+        if(_amount == 0) {
+            return _amount;
+        }
+        address[] memory path = new address[](2);
+        path[0] = address(weth);
+        path[1] = address(want);
+
+        uint256[] memory amounts = IUniswapV2Router02(router).getAmountsIn(_amount, path);
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: This should be getAmountsOut, we are trying to see how much "want" can we get out of exact amount token in

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+        uint256 toWithdraw = emissionToken.balanceOf(address(this)).sub(lpTokenBeforeBalance);
+
+        if (toWithdraw > 0) {
+            //if we have less than 18 decimals we need to lower the amount out
+            uint256 maxSlippage = toWithdraw.mul(DENOMINATOR.sub(slippageProtectionOut)).div(DENOMINATOR);
+            if(want_decimals < 18){
+                maxSlippage = maxSlippage.div(10 ** (uint256(uint8(18) - want_decimals)));
+            }
+
+            swapPool.removeLiquidityOneToken(toWithdraw, curveId, 0, now);
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol:387]: Missing moment where it wraps the native token after withdrawing it.

Eg.
`if (isWantWETH()) {
                weth.deposit{value: address(this).balance}();
            }`

File: avax-synapse-stablepool/contracts/Strategy.sol (hunk 1 above)
Comment (pajicf) [reply avax-synapse-stablepool/contracts/Strategy.sol:450]: missing weth wrapping edge case

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+    ///@notice This allows us to manually harvest with our keeper as needed
+    function setForceHarvestTriggerOnce(bool _forceHarvestTriggerOnce)
+        external
+        onlyAuthorized
+    {
+        forceHarvestTriggerOnce = _forceHarvestTriggerOnce;
+    }
+
+    ///@notice When our strategy has this much credit, harvestTrigger will be true.
+    function setMinHarvestCredit(uint256 _minHarvestCredit)
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: No need for this mechanism, we harvest on a time basis, this can make for more expensive bot cost

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+        external
+        onlyAuthorized
+    {
+        minHarvestCredit = _minHarvestCredit;
+    }
+
+    function getCurveId() external view returns (uint8) {
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: no need for this getter as curveId is already public

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+    function lpTokenToWant(uint256 tokens) public view returns (uint256) {
+        if (tokens == 0) {
+            return 0;
+        }
+
+        //we want to choose lower value of virtual price and amount we really get out
+        //this means we will always underestimate current assets. 
+        uint256 virtualOut = virtualPriceToWant().mul(tokens).div(1e18);
+
+        uint256 realOut = swapPool.calculateRemoveLiquidityOneToken(tokens, curveId);
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol:229]: Not sure if we really need this or if it affects accounting in some part of the system. @valo what do you think about this, the original strategy takes into account just the virtualOut

**Review by pajicf — CHANGES_REQUESTED** [review]
Some changes needed and left a open ended discussion for lpTokenToWant function

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 14]
```diff
@@ -0,0 +1,481 @@
...
+
+
+    uint8 private want_decimals;
+
+    uint8 public curveId;
+    address public targetToken;
+    bool public withdrawProtection;
+
+    bool internal forceHarvestTriggerOnce; // only set this to true externally when we want to trigger our keepers to harvest for us
+    uint256 public minHarvestCredit; // if we hit this amount of credit, harvest the strategy
```
Comment (pajicf) [inline avax-synapse-stablepool/contracts/Strategy.sol]: This should be removed as well

File: avax-synapse-stablepool/contracts/Strategy.sol (hunk 1 above)
Comment (pajicf) [reply avax-synapse-stablepool/contracts/Strategy.sol:450]: @joshualyguessennd

File: avax-synapse-stablepool/tests/test_setter.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    strategy.setForceHarvestTriggerOnce(False, {"from": gov})
+    tx = strategy.harvestTrigger(0, {"from": gov})
+    # print("\nShould we harvest? Should be false.", tx)
+    # assert tx == False
+
+    strategy.setMinHarvestCredit(0, {"from": gov})
+    tx = strategy.harvestTrigger(0, {"from": gov})
+    print("\nShould we harvest? Should be true.", tx)
+    assert tx == True
+    strategy.setMinHarvestCredit(2 ** 256 - 1, {"from": gov})
```
Comment (pajicf) [inline avax-synapse-stablepool/tests/test_setter.py]: This is excess now i guess..

**Review by pajicf — CHANGES_REQUESTED** [review]
I see a lot of TODOs in the tests. Can we make sure to implement them so we can also see what the APR is?

File: avax-synapse-stablepool/contracts/Strategy.sol (hunk 13 above)
Comment (valo) [reply avax-synapse-stablepool/contracts/Strategy.sol:229]: Seems like we should use one of these, not both. I think the only reason to use the virtualOut is for gas savings. Given that this is not a problem on sidechains, I think we should be find using only realOut. Unless there is a reason we need to take the min from both.

**Comment by joshualyguessennd** [comment]
Here is the original strategy
https://github.com/Grandthrax/SingleSidedCurve/blob/v5/043/contracts/Strategy.sol @valo

File: avax-synapse-stablepool/contracts/Strategy.sol (hunk 13 above)
Comment (valo) [reply avax-synapse-stablepool/contracts/Strategy.sol:229]: Actually the original strategy is using both, so there is probably a reason for this. Let's keep it as it is https://github.com/flashfish0x/SingleSidedCurve/blob/master/contracts/Strategy.sol#L103

File: avax-synapse-stablepool/contracts/Strategy.sol [hunk 16]
```diff
@@ -0,0 +1,483 @@
...
+    uint256 public slippageProtectionIn;// = 50; //out of 10000. 50 = 0.5%
+    uint256 public slippageProtectionOut;// = 50; //out of 10000. 50 = 0.5%
+    uint256 public constant DENOMINATOR = 10_000;
+    string internal strategyName;
+
+
+    uint8 private want_decimals;
+
+    uint8 public curveId;
+    address public targetToken;
```
Comment (valo) [inline avax-synapse-stablepool/contracts/Strategy.sol]: Why do we have `targetToken` while the original strategy is using `want`? I think `want` comes from `BaseStrategy` and adding a duplicate variable could lead to errors in case the 2 vars get out of sync.

**Comment by valo** [comment]
Is this strategy based on https://github.com/flashfish0x/SingleSidedCurve? I am seeing some differences, like the `_checkSlip` function, which is missing. I am not sure if I am comparing this strategy to the correct base strat.

**Comment by joshualyguessennd** [comment]
> Is this strategy based on https://github.com/flashfish0x/SingleSidedCurve? I am seeing some differences, like the `_checkSlip` function, which is missing. I am not sure if I am comparing this strategy to the correct base strat.

I used this one as reference, https://github.com/Grandthrax/SingleSidedCurve/blob/v5/043/contracts/Strategy.sol

File: avax-synapse-stablepool/contracts/Strategy.sol (hunk 13 above)
Comment (pajicf) [reply avax-synapse-stablepool/contracts/Strategy.sol:229]: Missed it, nevermind then

**Comment by pajicf** [comment]
Issues resolved, we still have that weird problem of strategy incurring loss unless a larger timespan between harvests happen. Let's put it on canary and see if any changes need to be made
